packages/react/src/combobox/item/ComboboxItem.tsx

```diff
@@ -0,0 +1,264 @@
+'use client';
+import * as React from 'react';
+import { useStore } from '@base-ui-components/utils/store';
+import { useLatestRef } from '@base-ui-components/utils/useLatestRef';
+import { useIsoLayoutEffect } from '@base-ui-components/utils/useIsoLayoutEffect';
+import { useAnimationFrame } from '@base-ui-components/utils/useAnimationFrame';
+
+import {
+  useComboboxRootContext,
+  useComboboxDerivedItemsContext,
+} from '../root/ComboboxRootContext';
+import {
+  useCompositeListItem,
+  IndexGuessBehavior,
+} from '../../composite/list/useCompositeListItem';
+import type { BaseUIComponentProps, HTMLProps } from '../../utils/types';
+import { useRenderElement } from '../../utils/useRenderElement';
+import { ComboboxItemContext } from './ComboboxItemContext';
+import { selectors } from '../store';
+import { useButton } from '../../use-button';
+import { useBaseUiId } from '../../utils/useBaseUiId';
+import { useComboboxRowContext } from '../row/ComboboxRowContext';
+
+/**
+ * An individual item in the popup.
+ * Renders a `<div>` element.
+ */
+export const ComboboxItem = React.memo(
+  React.forwardRef(function ComboboxItem(
+    componentProps: ComboboxItem.Props,
+    forwardedRef: React.ForwardedRef<HTMLDivElement>,
+  ) {
+    const {
+      render,
+      className,
+      value,
+      disabled = false,
+      nativeButton = false,
+      ...elementProps
+    } = componentProps;
+
+    const id = useBaseUiId();
+
+    const textRef = React.useRef<HTMLElement | null>(null);
+    const listItem = useCompositeListItem({
+      textRef,
+      indexGuessBehavior: IndexGuessBehavior.GuessFromOrder,
+    });
+
+    const {
+      store,
+      selectionMode,
+      getItemProps,
+      setOpen,
+      setSelectedValue,
+      valuesRef,
+      inputRef,
+      registerItemIndex,
+      allowActiveIndexSyncRef,
+      onItemHighlighted,
+      readOnly,
+      virtualized,
+      listRef,
+    } = useComboboxRootContext();
+
+    const { flatFilteredItems } = useComboboxDerivedItemsContext();
+
+    const selectable = selectionMode !== 'none';
+    const multiple = selectionMode === 'multiple';
+    const index = virtualized ? flatFilteredItems.indexOf(value) : listItem.index;
```

Comment:
Looks quite unefficient both ways. The index is a known property for a list item, it should be passed as a prop to the item.

Reply:
> to support an open API use case it needs to use the DOM order index

Or a required `index` property. What's the rationale for making `index` required on `Slider.Thumb` but not here? It solves the same underlying issue.